packages/eslint-plugin/src/rules/no-mixed-enums.ts

```diff
@@ -0,0 +1,222 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'tsutils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+type AllowedType = ts.TypeFlags.Number | ts.TypeFlags.String;
+
+export default util.createRule({
+  name: 'no-mixed-enums',
+  meta: {
+    docs: {
+      description: 'Disallow enums from having both number and string members',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      mixed: `Mixing number and string enums can be confusing.`,
+    },
+    schema: [],
+    type: 'problem',
+  },
+  defaultOptions: [],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const typeChecker = parserServices.program.getTypeChecker();
+
+    interface CollectedDefinitions {
+      imports: TSESTree.Node[];
+      previousSibling: TSESTree.TSEnumDeclaration | undefined;
+    }
+
+    function collectNodeDefinitions(
+      node: TSESTree.TSEnumDeclaration,
+    ): CollectedDefinitions {
+      const { name } = node.id;
+      const found: CollectedDefinitions = {
+        imports: [],
+        previousSibling: undefined,
+      };
+      let scope: Scope | null = context.getScope();
+
+      for (const definition of scope.upper?.set.get(name)?.defs ?? []) {
+        if (
+          definition.node.type === AST_NODE_TYPES.TSEnumDeclaration &&
+          definition.node.range[0] < node.range[0] &&
+          definition.node.members.length > 0
+        ) {
+          found.previousSibling = definition.node;
+          break;
+        }
+      }
+
+      while (scope) {
+        scope.set.get(name)?.defs?.forEach(definition => {
+          if (definition.type === DefinitionType.ImportBinding) {
+            found.imports.push(definition.node);
+          }
+        });
+
+        scope = scope.upper;
+      }
+
+      return found;
+    }
+
+    function getAllowedTypeForNode(node: ts.Node): AllowedType | undefined {
+      return tsutils.isTypeFlagSet(
+        typeChecker.getTypeAtLocation(node),
+        ts.TypeFlags.StringLike,
+      )
+        ? ts.TypeFlags.String
+        : ts.TypeFlags.Number;
+    }
+
+    function getTypeFromImported(
+      imported: TSESTree.Node,
+    ): AllowedType | undefined {
+      const type = typeChecker.getTypeAtLocation(
+        parserServices.esTreeNodeToTSNodeMap.get(imported),
+      );
+
+      const valueDeclaration = type.getSymbol()?.valueDeclaration;
+      if (
+        !valueDeclaration ||
+        !ts.isEnumDeclaration(valueDeclaration) ||
+        valueDeclaration.members.length === 0
+      ) {
+        return undefined;
+      }
+
+      return getAllowedTypeForNode(valueDeclaration.members[0]);
+    }
+
+    function getMemberType(
+      member: TSESTree.TSEnumMember,
+    ): AllowedType | ts.TypeFlags.Unknown | undefined {
+      if (!member.initializer) {
+        return undefined;
```

Comment:
why `undefined` here instead of just `Number`?

Reply:
Huh, nice spot, I guess it doesn't matter - all the tests pass either way.